File: kmm/server/src/main/kotlin/com/danioliveira/taskmanager/utils/UuidUtils.kt
package com.danioliveira.taskmanager.utils

import kotlin.time.Clock
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

/**
 * Extension function to generate a UUIDv7.
 * UUIDv7 is time-ordered, which provides better database indexing performance
 * compared to UUIDv4 (random).
 *
 * Benefits of UUIDv7:
 * - Time-ordered: Sequential nature improves B-tree index performance
 * - Sortable: Can be sorted by creation time
 * - Unique: Still maintains UUID uniqueness guarantees
 * - Database-friendly: Reduces index fragmentation
 */
@OptIn(ExperimentalUuidApi::class)
fun Uuid.Companion.randomV7(): Uuid {
    val value = random().toByteArray()
    val timestamp = Clock.System.now().toEpochMilliseconds()
    
    // Set version to 7 (0111 in binary)
    value[6] = (value[6].toInt() and 0x0F or 0x70).toByte()
    
    // Set variant to RFC 4122 (10 in binary)
    value[8] = (value[8].toInt() and 0x3F or 0x80).toByte()
    
    // Set the timestamp in the first 48 bits
    repeat(6) { value[it] = ((timestamp shr (40 - 8 * it)) and 0xFF).toByte() }
    
    return fromByteArray(value)
}

/**
 * Extension function to convert a String to Kotlin UUID.
 * Throws IllegalArgumentException if the string is not a valid UUID.
 */
@OptIn(ExperimentalUuidApi::class)
fun String.toUuid(): Uuid {
    return Uuid.parse(this)
}

File: task_manager_dart/server/lib/src/utils/uuid_utils.dart
import 'package:uuid/uuid.dart';

/// UUID utility class for generating UUIDs.
/// 
/// Uses UUIDv7 which is time-ordered, providing better database indexing performance
/// compared to UUIDv4 (random).
/// 
/// Benefits of UUIDv7:
/// - Time-ordered: Sequential nature improves B-tree index performance
/// - Sortable: Can be sorted by creation time
/// - Unique: Still maintains UUID uniqueness guarantees
/// - Database-friendly: Reduces index fragmentation
class UuidUtils {
  static const _uuid = Uuid();

  /// Generates a UUIDv7 (time-ordered UUID).
  /// 
  /// UUIDv7 embeds a timestamp in the first 48 bits, making it naturally
  /// sortable and providing better database performance than random UUIDs.
  static String generate() {
    return _uuid.v7();
  }

  /// Validates if a string is a valid UUID.
  static bool isValid(String uuid) {
    return Uuid.isValidUUID(fromString: uuid);
  }
}

